zebra-chain/src/keys/sprout.rs

```diff
@@ -1,14 +1,122 @@
-use std::{
-    fmt,
-    io::{self},
-};
+//! Sprout key types
+//!
+//! "The receiving key sk_enc, the incoming viewing key ivk = (apk,
+//! sk_enc), and the shielded payment address addr_pk = (a_pk, pk_enc) are
+//! derived from a_sk, as described in [‘Sprout Key Components’][ps]
+//!
+//! [ps]: https://zips.z.cash/protocol/protocol.pdf#sproutkeycomponents
+
+use std::fmt;
+
+use byteorder::{ByteOrder, LittleEndian};
 
 #[cfg(test)]
-use proptest::{array, collection::vec, prelude::*};
+use proptest::prelude::*;
 #[cfg(test)]
 use proptest_derive::Arbitrary;
 
-use crate::serialization::{SerializationError, ZcashDeserialize, ZcashSerialize};
+use sha2;
+
+/// Our root secret key of the Sprout key derivation tree.
+///
+/// All other Sprout key types derive from the SpendingKey value.
+/// Actually 252 bits.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct SpendingKey(pub [u8; 32]);
+
+/// Derived from a _SpendingKey_.
+pub type ReceivingKey = x25519_dalek::StaticSecret;
+
+impl From<SpendingKey> for ReceivingKey {
+    /// For this invocation of SHA256Compress as the PRF, t=0, which
+    /// is populated by default in an empty block of all zeros to
+    /// start.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#sproutkeycomponents
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteprfs
+    fn from(spending_key: SpendingKey) -> ReceivingKey {
+        let mut state = [0u32; 8];
+        let mut block = [0u8; 64]; // Thus, t = 0
+
+        block[0..32].copy_from_slice(&spending_key.0[..]);
```

Comment:
The type of `SpendingKey` doesn't enforce (or document) that the 4 high-order bits of the first byte must be zeros. Therefore this could end up copying 1 into the bits of weight 2^4 or 2^5 of `block[0]`, which would violate domain separation.

Reply:
Also document that this is implementing PRF^addr (or factor it into a separate function named to make that obvious).

Reply:
I pushed some updates to how we can construct these that should be better?

Reply:
This looks great, thanks. Add a comment here or an assertion that the top 4 bits are zero.